Implement a simple node.js cli tool.

junseinagao opened the pull request:
I've implemented a simple cli on work with npm.

I think budoux is a great tool for Japanese web development and people need CLI tool using more easier.  people can format texts only to install Node.js and run `npx budoux-cli`.

For test locally.
```bash
$ cd node_cli
$ npm link
$ budoux-cli -H 
```

Output Example.
```bash
$budoux-cli
Please, pass one text argument to translate at least.
$budoux-cli --help
usage: budoux [-h] [-H] [-m JSON] [-d STR] [-V] [TXT]

オプション:
  -H, --html     HTML mode                                                [真偽]
      --version  バージョンを表示                                         [真偽]
      --help     ヘルプを表示                                             [真偽]
$budoux-cli --version
0.0.1
$budoux-cli 今日は天気です。
今日は
天気です。
$budoux-cli '今日は天気です。'
今日は
天気です。
$budoux-cli -H '今日は<b>とても天気</b>です。'
<span style="word-break: keep-all; overflow-wrap: break-word;">今日は<b><wbr>とても<wbr>天気</b>です。</span>
```

eggplants commented:
Are `-m JSON`, `-d STR`, `-V` options not yet  implemented?

eggplants commented on node_cli/tests/test_main.ts:
```diff
@@ -0,0 +1,118 @@
+  it('should output the help when exec command with --help option.', () => {
+    const expectedText = `usage: budoux [-h] [-H] [-m JSON] [-d STR] [-V] [TXT]
+
+オプション:
+  -H, --html     HTML mode                                                [真偽]
+  -d, --delim    output delimiter in TEXT mode                          [文字列]
+  -m, --model    custom model file path or json string                  [文字列]
+  -V, --version  バージョンを表示                                         [真偽]
+  -h, --help     ヘルプを表示                                             [真偽]
+`;
```
These Japanese text generated by [yargs](https://github.com/yargs/yargs) like `オプション`, `真偽` and `バージョンを表示` are depended on your locale. On other environment, especially on CI, these maybe will be changed into English ones and tests will fail. So you have to change testcases or disable the i18n of yargs.

junseinagao replied:
Thank you for your review. I'm going to fix it.

tushuhei replied:
I think this test case can be removed since it looks like a test more for yargs.

eggplants commented on node_cli/tests/test_main.ts:
```diff
@@ -0,0 +1,118 @@
+  it('should output the help when exec command with -h option alias.', () => {
+    const expectedText = `usage: budoux [-h] [-H] [-m JSON] [-d STR] [-V] [TXT]
+
+オプション:
+  -H, --html     HTML mode                                                [真偽]
+  -d, --delim    output delimiter in TEXT mode                          [文字列]
+  -m, --model    custom model file path or json string                  [文字列]
+  -V, --version  バージョンを表示                                         [真偽]
+  -h, --help     ヘルプを表示                                             [真偽]
+`;
```
Same.

junseinagao commented:
I'm stucking in this implementation.

`budoux-cli -d '---' 今日は天気です。` is going to be work well but `budoux-cli --delim '---' 今日は天気です。` doesn't work.

It may a something bug in `yargs`. Anyway, it will take me more time to fix this problem.

tushuhei commented:
Hi there, thank you for your proposal. I believe it will be useful for those who want to use the CLI without Python environment. That being said, I want to make it behave the same as the one written in Python by @eggplants to avoid the confusion between the Python version and the JavaScript version.

Could you consider to solve items below for consistency?
- Please consider to move the CLI code under the `javascript` directory and add the "bin" field to the existing package.json. I want to avoid making another NPM package to maintain for the CLI functionality.
- The command name could be simplified as "budoux".
- I feel the biggest difference between the proposed CLI and the existing one is the support for standard inputs (e.g. use cases such as `echo "今日は天気です。" | budoux`). It would be great if the JavaScript version also supports that functionality.
- I believe the meaning of the `delimiter` option differs from the existing one. It's the character to separate multiple text inputs concatenated by line breaks.
```
$ echo -e "今日は元気です\n明日は雨かな？" | budoux
今日は
元気です
---  <-- this is a delimiter.
明日は
雨かな？
```

tushuhei reviewed (changes requested):
Great work! Some minor fixes needed.

tushuhei commented:
@junseinagao Also, please visit https://cla.developers.google.com/ and complete signing CLA.
```
If you are not currently covered under a CLA, please visit https://cla.developers.google.com/. Once you've signed (or fixed any issues), please request a rescan from a repository manager.
```

tushuhei commented on javascript/package.json:
```diff
@@ -9,6 +9,9 @@
   "main": "./dist/index.js",
   "module": "./module/index.js",
+  "bin": {
+    "budoux": "bin/budoux.js"
```
Please update the path to "./bin/budoux.js" for consistency.

junseinagao commented:
@googlebot I signed.

eggplants commented:
Did you specify the arity of TXT in parser?
My implementation in Python restricts TXT's arity to only one.

tushuhei commented:
Approving the change since my requests are reflected properly.

As @eggplants pointed out, it would be great if items below can be solved in the following changes.

- Inconsistency in the number of arguments accepted for TXT
```
$ budoux 今日は寒いですね。 明日はどうでしょう？
usage: budoux [-h] [-H] [-m JSON] [-d STR] [-V] [TXT]
budoux: error: unrecognized arguments: 明日はどうでしょう？
```

```
```

tushuhei reviewed (changes requested):
Actually, I found that optional arguments also accept more than one values, which makes some of the test cases fail. Please fix to pass the unit test.

junseinagao commented:
Okey, I'm going to fix these problems.

junseinagao commented:
@tushuhei 

I fixed the program and here is changed points.

### get the only one argument and stdin mode.

The program get 0 or 1 argument. (commander.js support this behavior to use placeholder `[txt]` instead of `<txt>`.) If the program get 0 argument, a user can input stdin and program parse stdin. If the program get more than one argument, throw error.

### Add some tests.

I added some tests.

- the program success when paese `--model` option with sample custom model.
- the program success when parse `--version` or `-V` option.
- the program fail when the program paese `-v`.
- the program fail when get the extra program-arugments.
- the program fail when get the extra option-arguments.

### Maybe fix that optional arguments also accept more than one values.

Actually, I haven't reproduced the error you said.

I believe options (`--delim` and `--model`) won't get option-argument more than one value and parse its as a program-argument. Now, the program won't get more than one program-argument and throw error if get extra option-argument.

Please, tell me the specific failed test If you get tests to failed yet.

### Import version number from `package.json` but reverted.

I tried to import version number from `package.json` but it occur change the directory structure of `dist` and `module` so I don't implement it's function and I only add the test that version is the same as package.json.

eggplants commented:
@tushuhei could you push `Approve and run` to run test CI?

eggplants commented:
@tushuhei Sorry for my mistake in CI. Could you re-run?

tushuhei commented:
Hmm, I'm not sure how I can rerun the test with the latest configuration including `npm link`. It seems running on the old one.

eggplants commented:
@junseinagao please try:
`git commit --allow-empty -m "empty" && git push`

junseinagao commented:
Pushed. I wish to pass tests in CI.

eggplants commented:
@tushuhei Please `Approve and run` again🙏

tushuhei commented:
It seems the test is failing only on Windows. I'm not 100% sure, but the causation might be around the execution policy.
https://go.microsoft.com/fwlink/?LinkID=135170

@junseinagao Could you update your test code to run `npx budoux` instead of `budoux`? `npx` should run as long as Node.js is installed, so I'm curious to see if this solves the issue.

tushuhei commented:
Probably the safest way would be not relying on `execFile` entirely. The problem should not occur as long as we test the `cli` method itself. I suggest to rewrite the test only to use the `cli` method as you do with the case: `should output the error message when get more than one text argument.`.

junseinagao commented:
@tushuhei 

Just now, I'm thinking about that point. Testing with `execFile` has a lot of problem.

Okey, I'm going to rewrite tests to import `cli` methods.

(Under text is my writing text in the middle.)

### `npx` don't parse EOL in argument'

This is my local zsh output. This behavior is a kind of mysterious. I guess both of `budoux` and `npx budoux` call `node bin/budoux.js` 🙃.

```zsh
junsei@mbp2019 javascript % budoux '今日は元気です。
quote> 明日は晴れかな？'
今日は
元気です。
---
明日は
晴れかな？
junsei@mbp2019 javascript % npx budoux '今日は元気です。
明日は晴れかな？'
今日は
元気です。
\n明日は
晴れかな？
```

eggplants commented:
Shall we merge this PR once and create another PR with Windows support? It looks like this ​works perfectly except for the testing...

tushuhei commented:
Sounds good. Let's merge this PR for now and start a new issue for the Windows support.

junseinagao commented:
Thanks for merging PR! 😊 I will request a new PR for the WIndows support soon.
